src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIMarsConnection.cs

```diff
@@ -219,228 +192,250 @@ public void HandleSendComplete(SNIPacket packet, uint sniErrorCode)
+                        switch (_state)
                         {
-                            currentHeader = null;
-                            currentPacket = null;
-                            currentSession = null;
-
-                            while (_currentHeaderByteCount != SNISMUXHeader.HEADER_LENGTH)
-                            {
-                                int bytesTaken = packet.TakeData(_headerBytes, _currentHeaderByteCount, SNISMUXHeader.HEADER_LENGTH - _currentHeaderByteCount);
-                                _currentHeaderByteCount += bytesTaken;
+                            case State.Header:
+                                int taken = packet.TakeData(_headerBytes, _headerCount, SNISMUXHeader.HEADER_LENGTH - _headerCount);
+                                _headerCount += taken;
+                                SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNIMarsConnection), EventType.INFO, "MARS Session Id {0}, took {1} header bytes", args0: _lowerHandle?.ConnectionId, args1: packet.DataLeft,args2: taken);
+                                if (_headerCount == SNISMUXHeader.HEADER_LENGTH)
+                                {
+                                    _header.Read(_headerBytes);
+                                    _payloadLength = (int)_header.length;
+                                    _payloadCount = 0;
+                                    _partial = RentPacket(headerSize: 0, dataSize: _payloadLength);
+                                    _state = State.Payload;
+                                    SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNIMarsConnection), EventType.INFO, "MARS Session Id {0}, header complete, _payloadLength {1}", args0: _lowerHandle?.ConnectionId, args1: _payloadLength);
+                                    goto case State.Payload;
+                                }
+                                else
+                                {
+                                    loopState = LoopState.Recieve;
+                                }
+                                break;
 
-                                if (bytesTaken == 0)
+                            case State.Payload:
+                                if (packet.DataLeft == _payloadLength && _partial == null)
                                 {
-                                    sniErrorCode = ReceiveAsync(ref packet);
-                                    SqlClientEventSource.Log.TrySNITraceEvent(s_className, EventType.INFO, "MARS Session Id {0}, Non-SMUX Header SNI Packet received with code {1}", args0: ConnectionId, args1: sniErrorCode);
+                                    // if the data in the packet being processed is exactly and only the data that is going to sent
+                                    // on to the parser then don't copy it to a new packet just forward the current packet once we've
+                                    // fiddled the data pointer so that it skips the header data when
```

Comment:
Don't leave me in suspense. When what?

Reply:
When you know, the thing, that thing, happens....
I think originally it was that that partial packet was null but it's just a trailing off now. Removed, The rest of the comment makes sense without that word I think but I'll left you judge if the explanations make sense.